### ISSUE TITLE
Running multiple `cf set-env` commands in parallel has inconsistent results

### ISSUE THREAD

**Issue by dmikusa-pivotal** [cloudfoundry/cli#417]
If you run multiple `cf set-env` commands at the same time, the results are inconsistent and never correct.

Here's an example script to replicate this:

```bash
#!/bin/bash
export CF_TRACE=trace.log
counter=1
while [ $counter -le 4 ]
do
    envName="testEnv$counter"
    cf set-env $1 $envName $envName &
    counter=$((counter+1))
done
wait
echo All done
```

The expected result would be to have four environment variables set like "testEnv1" through "testEnv4".  What happens in practice is that you end up with one or two of the variables set.

Looking at the trace logs, you can see what is happening.  The command first sends a GET request to pull the current list of environment variables and then sends a PUT request to update it with the new environment variable.  Each separate `cf set-env` process runs the GET and sees the initial state of the environment variables.  It then runs the PUT with a single new variable added.  Unfortunately which ever one runs last wins and steps on the other changes.

The use case for this is when there are a large number of environment variables being set.  Running multiple `cf set-env` commands in parallel can speed things up.

**Comment by dmikusa-pivotal** [comment]
Currently you can work around this by using a manifest.yml to specify the environment variables.  Not everyone uses a manifest though, so it would be nice to be able to do something similar with cli arguments.

**Comment by goehmen** [comment]
The CLI team is working on concurrency issues after noting that the introduction of plugins enables concurrency in the CLI like never before.  Thanks for the issue/tracker story.

**Comment by goehmen** [comment]
Plugin concurrency didn't address this issue.  The problem is that there isn't a better way to handle this short of atomic transactions but atomic transactions is not an available solution.  Runtime would have to provide that atomicity.

**Comment by dmikusa-pivotal** [comment]
@goehmen - would you consider a feature request to allow setting multiple env variables via the command line?  Something like `cf set-env some-great-app env1 val1 env2 val2 env3 val3 ...`.

While this doesn't address the problem listed here, it does seem to address the core desire of the user, which is to quickly set multiple env variables without a manifest.

**Comment by maximilien** [comment]
Maybe a better syntax if you consider this suggest to make parsing easier:

```
$ cf set-env some-great-app env1=val1, env2=val2, env3="val3" ...
```

using `"..."` for values with spaces.

**Comment by dmikusa-pivotal** [comment]
+1 I like the `env1=val1` style too.  A couple comments...

 - Currently env variable names can have spaces too.  Would "some env"="some val" work? or "some env=some val"?

- Can we drop the comma after each pair?  `env1=val1 env2=val2 ...`

**Comment by maximilien** [comment]
@dmikusa-pivotal +1 on your update

**Comment by goehmen** [comment]
I like it too.  I updated the tracker story with this suggestion.

Greg Oehmen
Cloud Foundry Product Manager
415.205.6596

**Comment by fny** [comment]
This would be fantastic. Both Heroku and AWS Elastic Beanstalk use a similar syntax:

    heroku config:set KEY1=VALUE1 [KEY2=VALUE2 ...]
    eb setenv HeapSize=256m Site_Url=mysite.elasticbeanstalk.com

To use spaces, the values can be quoted (e.g. `KEY="SOME VALUE"` or `KEY='SOME VALUE'`)

**Comment by tanob** [comment]
What's the status on this one? I can't see the story in the pivotal tracker but I see `set-env` still doesn't support setting multiple variables.

**Comment by Omnipresent** [comment]
Is there any status on this? What is a short term fix? I like the suggestions of `env1=val1 env2=val2` but for the short term would multiple commands with `&&` work? i.e. 

```
cf set-env someapp env1 val1 && cf set-env someapp env2 val2
```

**Comment by dmikusa-pivotal** [comment]
You could try `cf curl -X PUT /v2/apps/<app-guid> -d '{"environment_json":{"ENV1":"VAL1", "ENV2": "VAL2"}}'`.  That overwrites everything and replaces it with what is in the JSON, so if you have existing env variables you'd have to include those as well as the new ones you want to set.

**Comment by Omnipresent** [comment]
@dmikusa-pivotal Thanks. that is a good find. However, this seems like an overkill since everything in manifest.yml is overwritten. I have several env variables in manifest.yml that I would hate to overwrite again via the `PUT`. The reason I want to set some env variables via the CLI is because they are passwords and I don't want to store them in manifest.yml in clear text and have them be part of my code and repository.

**Comment by dkoper** [comment]
@Omnipresent Multiple commands with `&&` should work, but that just runs them one by one, not addressing the bad performance reported in the issue description.

The story in our Pivotal tracker seems to have been deleted -  I have removed the comment with the link to it.

Note that running cf commands in parallel (like with `&`) could lead to unpredictable issues: when the access token expires, that command process will retrieve a new token from UAA and save it to your local config.json: any command running in parallel, depending on timing, could be reading from an incomplete file as it is being written to. So if you must do that, please use a different CF_HOME for each parallel stream.

I don't think my team will get to working on this feature any reasonable time soon; too many things on our plates.
In the meantime a possible work-around that has helped some users is to take advantage of app manifest inheritance: move your environment variables to a "parent" manifest, which you can store in a secure location (separate to your app in your repo), or create on the fly.
When your app manifest inherits this file, these environment variables will be applied to all apps declared within. (So this may not be an ideal solution if you have a manifest with multiple apps.)

**Comment by dkoper** [comment]
FYI, I noticed in another issue someone pointing out that maybe secrets should be passed to apps using user-provided services, configured before the app push, possibly by a different person/role (the person responsible for the resource the secret is used to access).

**Comment by jgriff** [comment]
This seems to also be an issue for those using the [cloudfoundry-client-reactor](https://github.com/cloudfoundry/cf-java-client) client lib to set env variables asynchronously (ie, multiple variables set in parallel for same app).  

For example, each [`Applications#setEnvironmentVariable(SetEnvironmentVariableApplicationRequest)`](https://github.com/cloudfoundry/cf-java-client/blob/master/cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/Applications.java#L203) request sets only 1 environment variable, and when multiple are executed in parallel you have the same problem.

Was hoping there was a way to set multiple env vars in a single request, but I cannot find it.  Only workaround at this point seems to be to line up the requests in serial where each request waits for the previous one to complete before executing the next.  But this obviously negates much of the benefit of the reactive client.

Any suggestions?

**Comment by dmikusa-pivotal** [comment]
@jgriff - I don't know about the library you're referencing, but the cf cli sets multiple variables at once by doing a PUT against `/v2/apps/<guid>`. There is an `environment_json` attribute which can be used to set multiple at once.

http://apidocs.cloudfoundry.org/5.0.0/apps/updating_an_app.html
